spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala

```diff
@@ -0,0 +1,1053 @@
+    trainingModel: Module[T])(implicit ev: TensorNumeric[T]): Unit = {
+    cacheTrigger.foreach { trigger =>
+      cachePath.foreach { path =>
+        if (trigger(state)) {
+          saveModel(getModel(models, trainingModel), cachePath, isOverWrite,
+            s".${state[Int]("neval")}")
+          logger.info(s"[Wall Clock ${wallClockTime / 1e9}s] Save model to $path")
+          optimMethods.foreach{case (name, optimMethod) =>
+            optimMethod.state.update("epoch", state[Int]("epoch"))
+            optimMethod.state.update("neval", state[Int]("neval"))
+            saveOptimMethod(optimMethod, cachePath, isOverWrite, s"-$name.${state[Int]("neval")}")
+            logger.info(s"[Wall Clock ${wallClockTime / 1e9}s] Save optimMethod " +
+              s"${optimMethod} to $path")
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Save train summaries.
+   *
+   * @param trainSummary train logger
+   * @param models cached models
+   * @param driverState driver state
+   */
+  private def saveSummary[T: ClassTag](
+    trainSummary: TrainSummary,
+    models: RDD[Cache[T]],
+    driverState: Table,
+    trainingModel: Module[T])(implicit ev: TensorNumeric[T]): Unit = {
+    val currentIteration = driverState[Int]("neval") - 1
+    val parametersTrigger = trainSummary.getSummaryTrigger("Parameters")
+    if (parametersTrigger.isDefined && parametersTrigger.get(driverState)) {
+      val model = getModel(models, trainingModel)
+      val parametersTable = model.getParametersTable()
+      // Parallelize to create Histogram.
+      Engine.default.invokeAndWait(
+        parametersTable.keySet.toSeq.map(moduleName => () => {
+          val paramTable = parametersTable[Table](moduleName)
+          paramTable.keySet.foreach { paramName =>
+            trainSummary.addHistogram(
+              s"$moduleName/$paramName", paramTable[Tensor[T]](paramName), currentIteration)}
+        }))
+    }
+    val scalarTrigger = trainSummary.getScalarTriggers()
+    // Not parallelizable, because driverState is changing each iteration.
+    scalarTrigger.foreach { v =>
+      if (v._2(driverState)) {
+        require(driverState.contains(v._1), s"ParallelOptimizer.saveSummary: Summary ${v._1} " +
+          s"is not supported now.")
+        trainSummary.addScalar(
+          v._1, driverState[Float](v._1), currentIteration
+        )
+      }
+    }
+  }
+
+  /**
+   * Init engine and cache models, weights, gradients, criterions, state tables
+   * and validation methods on worker nodes.
+   *
+   * @param model train model
+   * @param dataset train dataset
+   * @param criterion loss function
+   * @param state state table
+   * @param nodeNumber node number
+   * @param coresPerNode cores per node
+   * @param checkSingleton if checkSingleton
+   * @param validationMethods validation methods
+   * @param optimMethod optimization method
+   * @return cached models
+   */
+  private def initThreadModels[T: ClassTag](
```

Comment:
same with above

Reply:
this method behaviours quite different from the original one (setting synchronizer, etc), I'll check if we can do some consolidation or just leave it